### DISCUSSION TITLE
How "update" FutureProvider?

### DISCUSSION THREAD

**Discussion by FetFrumos** [rrousselGit/riverpod#2256] [General]
I have FutureProvider in my app:

     ```
    final myFutureProvider = FutureProvider<ProductLoadResult<Data>>((ref) async {
    final connectivityResult = await (Connectivity().checkConnectivity());
    final netAvailable = connectivityResult != ConnectivityResult.none;
    if(netAvailable){
    final productsApi = ref.read(productsApiProvider);
    final result = await productsApi.getData();
      return ProductLoadResult<Data>(true, result);
    }
    else{
      return ProductLoadResult<Data>(false, null);
     }
   });
   ```
   
 therefore, in this provider, I check the network status and if the network is available, I receive data via the API.

 using the provider:
   
   
          ```
             final apiData =  ref.watch(myFutureProvider);
           
              body: apiData.when(
                  data: _bodyForData,
                  error: _bodyForError,
                 loading: () => const Loader()),
         ```
                 
if the network is not available, I display the button. on pressure the following logic:    
  
      ```
      setState(() {
              //without this line, the value of categoriesFutureProvider is not updated
              ref.invalidate(categoriesFutureProvider);
        });
      ```
        
it works - but I'm not sure it's the optimal logic.

maybe there is a better option? I will be grateful for your help

**Comment by karlhp** [comment]
"if the network is not available, I display the button. on pressure the following logic: "

I wonder why you need setState() in this place. Did you try the following?

```
TextButton(
                onPressed: () => ref.invalidate(categoriesFutureProvider)
                child: Text('Retry'),
              ),
```

The other part looks okay, I would propably place the connectivity check in the API or HTTP Client repository and throw a custom exception if there is no-connectivity. Then you would be able to handle the no-connectivity and any other exception in "error: _bodyForError", so you wouldn't need the "if (netAvailable) ...." statement in this and other providers with API requests.
